test/libsolutil/FixedHash.cpp

```diff
@@ -0,0 +1,270 @@
+	FixedHash<8> e("112233445566778899");
+	BOOST_CHECK_EQUAL(e.size, 8);
+	BOOST_CHECK_EQUAL(e.hex(), "0000000000000000");
+}
+
+BOOST_AUTO_TEST_CASE(string_constructor_frombytes)
+{
+
+	FixedHash<8> b("", FixedHash<8>::FromBinary);
+	BOOST_CHECK_EQUAL(b.size, 8);
+	BOOST_CHECK_EQUAL(b.hex(), "0000000000000000");
+
+	FixedHash<8> c("abcdefgh", FixedHash<8>::FromBinary);
+	BOOST_CHECK_EQUAL(c.size, 8);
+	BOOST_CHECK_EQUAL(c.hex(), "6162636465666768");
+
+	// TODO: short input, this should fail
+	FixedHash<8> d("abcdefg", FixedHash<8>::FromBinary);
+	BOOST_CHECK_EQUAL(d.size, 8);
+	BOOST_CHECK_EQUAL(d.hex(), "0000000000000000");
+
+	// TODO: oversized input, this should fail
+	FixedHash<8> e("abcdefghi", FixedHash<8>::FromBinary);
+	BOOST_CHECK_EQUAL(e.size, 8);
+	BOOST_CHECK_EQUAL(e.hex(), "0000000000000000");
+}
+
+BOOST_AUTO_TEST_CASE(converting_constructor)
+{
+	// Truncation
+	FixedHash<8> a = FixedHash<8>(FixedHash<12>("112233445566778899001122"));
+	BOOST_CHECK_EQUAL(a.size, 8);
+	BOOST_CHECK_EQUAL(a.hex(), "1122334455667788");
+
+	// Left-aligned extension
+	FixedHash<12> b = FixedHash<12>(FixedHash<8>("1122334455667788"), FixedHash<12>::AlignLeft);
+	BOOST_CHECK_EQUAL(b.size, 12);
+	BOOST_CHECK_EQUAL(b.hex(), "112233445566778800000000");
+
+	// Right-aligned extension
+	FixedHash<12> c = FixedHash<12>(FixedHash<8>("112233445566778899001122"), FixedHash<12>::AlignRight);
+	BOOST_CHECK_EQUAL(c.size, 12);
+	// TODO: Why is this failing?
+	BOOST_CHECK_EQUAL(c.hex(), "000000000000000000000000");
+
+	// Default setting
+	FixedHash<12> d = FixedHash<12>(FixedHash<8>("1122334455667788"));
+	BOOST_CHECK_EQUAL(d, b);
+
+	// FailIfDifferent setting
+	// TODO: Shouldn't this throw?
+	FixedHash<12> e = FixedHash<12>(FixedHash<8>("1122334455667788"), FixedHash<12>::FailIfDifferent);
+	BOOST_CHECK_EQUAL(e, b);
+}
+
+BOOST_AUTO_TEST_CASE(arith_constructor)
+{
+	FixedHash<20> a(u160(0x1234));
+	BOOST_CHECK_EQUAL(
+		a.hex(),
+		"0000000000000000000000000000000000001234"
+	);
+
+	FixedHash<32> b(u256(0x12340000));
+	BOOST_CHECK_EQUAL(
+		b.hex(),
+		"0000000000000000000000000000000000000000000000000000000012340000"
+	);
+
+	// NOTE: size-mismatched constructor is not available
+}
+
+BOOST_AUTO_TEST_CASE(to_arith)
+{
+	FixedHash<20> a{};
+	BOOST_CHECK_EQUAL(u160(a), 0);
+
+	FixedHash<32> b("c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470");
+	BOOST_CHECK_EQUAL(u256(b), u256("0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470"));
+}
+
+BOOST_AUTO_TEST_CASE(comparison)
+{
+	FixedHash<32> a("c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470");
+	FixedHash<32> b("c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470");
+	FixedHash<32> c("c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a471");
+	FixedHash<32> d("233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470c5d2460186f7");
+
+	BOOST_CHECK(a == a);
+	BOOST_CHECK(b == b);
+	BOOST_CHECK(a == b);
+	BOOST_CHECK(a != c);
+	BOOST_CHECK(a != d);
```

Comment:
For completeness, perhaps `b == a` and `c != a` as well? Fine without it though.

Reply:
Different specialisations can't be compared, will leave a comment.

Reply:
Ah you just meant the opposite order, sure.